Inspect supports alias

gpoirier opened the pull request:
- [x] I've read the [CONTRIBUTING](https://github.com/sbt/sbt/blob/1.x/CONTRIBUTING.md) guidelines

Fixes https://github.com/sbt/sbt/issues/2881

Commit:
Adding minimal support for commands in inspect

There's also a special case for aliases that will try to resolve
the target of the alias to a task key if possible and display the
output of that key if found.

see https://github.com/sbt/sbt/issues/2881

gpoirier commented:
I signed the CLA

eed3si9n commented:
@gpoirier Thanks for the contribution!

eed3si9n reviewed (changes requested):
Could you update `help inspect` in https://github.com/sbt/sbt/blob/1.x/main/src/main/scala/sbt/internal/CommandStrings.scala#L122-L123 too?

gpoirier commented on main/src/main/scala/sbt/internal/Inspect.scala:
```diff
@@ -27,19 +27,18 @@ object Inspect {
     }
-  val spacedModeParser: (State => Parser[Mode]) = (s: State) => {
+  val spacedModeParser: State => Parser[Mode] = (_: State) => {
+    val default = "-" ^^^ Details(false)
```
@eed3si9n @dwijnand Is it a decent solution to wanting inspect the key named 'actual' with `Details(false)`? I can leave it out or renamed `-` to a different name if preferred.

eed3si9n replied:
It's a bit ironic since `-` is an actual command name in sbt 0.13 to do early processing, like `-warn`, but I think `inspect - somekey` is a decent escaping solution.

gpoirier commented:
Good point, I missed the `help inspect` message. I'll update it, and I should also add something about the `-` mode if you guys want that in? https://github.com/gpoirier/sbt/blob/issues/2881/main/src/main/scala/sbt/internal/Inspect.scala#L36

I'll also have to figure out how to run the scripted tests locally so I don't rely on travis to know if my fix worked - because my first attempt didn't.

gpoirier commented:
It seems like this PR should be linked to the issue https://github.com/sbt/sbt/issues/2881, but I cannot figure out how to do that.

gpoirier commented:
I updated the help text, does it seems appropriate?

eed3si9n commented:
@gpoirier 

The beginning of the help text says:


>     |	For a plain setting, the value bound to the key argument is displayed using its toString method.
>     |	Otherwise, the type of task ("Task" or "Input task") is displayed.

Maybe it should be updated a bit to reflect your main change. For example:

```
For a plain setting, the value bound to the key argument is displayed using its toString method.
For an alias, the command bound to the alias is displayed.
Otherwise, the type of the key ("Task" or "Input task") is displayed.
```

gpoirier commented:
With the help change above, is it instead of the text I added at the bottom? The main reason I added it at the bottom was that the change apply to all of the modes.

eed3si9n commented:
Either works for me in terms of the text.

eed3si9n commented:
Overall LGTM.
Could you squash the commits please?

gpoirier commented:
It's squashed to a singe commit rebased on the current HEAD of 1.x.

eed3si9n commented:
Awesome.

eed3si9n commented:
I am testing it locally, and seeing something like this:

```
sbt:helloworld> inspect show
[error] Not a valid project ID: show
[error] Expected ':'
[error] Not a valid key: show (similar: showTiming, sLog, ps)
[error] inspect show
[error]             ^
sbt:helloworld> inspect +
[error] Expected whitespace character
[error] Expected '-'
[error] Expected 'tree'
...
```

Is this the expected result for commands?

eed3si9n reviewed (changes requested):
See comment

gpoirier commented:
Those two (`+` and `show`) aren't in the list from `State#definedCommands`, which is what @dwijnand asked me to use. As far as I can tell, `show` doesn't seem to be technically be a command, or at least I can't find it by browsing the source if it's the case. Also, I can only use the `SimpleCommand`s, the `ArbitraryCommand`s doesn't expose a name.

The two error messages you listed are also different for reasons that already existed. One of the parser for the keys is looking at IDs before trying to match the result to a key. Which I'm realizing right now might be an issue, since technically keys could have names that are not valid IDs. But this PR isn't changing that behavior.

gpoirier commented:
@eed3si9n do you still want me to change something? If so, what should it be?

eed3si9n reviewed (approved):
ok. Let's merge this as is. Since adding support for alias was the original goal, so this PR does the job.

gpoirier commented:
Should a new ticket be open for the part that's not related to adding alias support then? What was your issue though, the fact that `show` and `+` weren't seen as commands or the error message? If it's the former, I don't think it can easily be solved.

eed3si9n commented:
I guess I was hoping that inspect somehow recognizes `+` as a command. I think it's ok as is.

We made a survey form to collect feedback on the contribution process. Please fill it out when get you get a minute. https://goo.gl/forms/LDPlTl9Wso7YwEri2
